preload/main.c

```diff
@@ -22,28 +23,40 @@ int ioctl (int d, int request, char *argp) {
     if (orig_ioctl == NULL) {
         orig_ioctl = dlsym(RTLD_NEXT, "ioctl");
     }
+
+    /* We only care for TIOCGWINSZ ioctls */
+    if (request != 0x5413) {
+        return orig_ioctl(d, request, argp);
+    }
+
     static int max_rows = -1;
+
+    /* ioctl gets called once before the perl module had the time to determine
+     * the right size! Leave max_rows negative to indicat that it still needs to
+     * be read from the SHELLEX_SIZE_FILE */
+
     if (max_rows < 0 ) {
-        char *str = getenv("SHELLEX_MAX_ROWS");
+
+      char *fname = getenv("SHELLEX_SIZE_FILE");
+      FILE *stream = fopen (fname,"r");
+      char str[5] = "-500";
+      if (stream) {
+        char *ret = fgets(str,6,stream);
+        fclose(stream);
         if (str != NULL) {
-            max_rows = atoi(str);
+          /* this may be -500 */
+          max_rows = atoi(str);
         }
+      }
     }
 
-    // We only care for TIOCSWINSZ ioctls
-    if (request != 0x5414) {
-        return orig_ioctl(d, request, argp);
-    }
+    int retval = orig_ioctl(d, request, (char *)argp);
+    struct winsize *ws = (struct winsize *)argp;
 
-    struct winsize ws = *((struct winsize *)argp);
-    int fheight = ws.ws_ypixel / ws.ws_row;
-    if (max_rows < 0) {
-        ws.ws_row = 80;
-        ws.ws_ypixel += 80 * fheight;
-    } else {
-        ws.ws_row = max_rows;
-        ws.ws_ypixel += max_rows * fheight;
-    }
+    /* max_rows is still negative at first invocation */
+    int fheight = ws->ws_ypixel / ws->ws_row;
+    ws->ws_row = (max_rows>0)?max_rows:25;
+    ws->ws_ypixel = ws->ws_row * fheight;
 
-    return orig_ioctl(d, request, (char *)&ws);
+    return retval;
```

Comment:
The manipulation of ws is actually used, just not inside the function. ws is just a typecast of argp, which is a pointer from where ioctl is called.